### PR TITLE
Fix for batching when the field type is a list

### PR DESCRIPTION
**Pull request by sujeetsr**
**Summary:**  
Fix for batching when the field type is a list

**Issue Reference(s):**  
Fixes #... _(Replace "..." with the issue number)_

**Build & Testing:**

- [x] I ran `cargo test` successfully.
- [x] I have run `./lint.sh --mode=fix` to fix all linting issues raised by `./lint.sh --mode=check`.

**Checklist:**

- [x] I have added relevant unit & integration tests.
- [x] I have updated the documentation accordingly (if applicable).
- [x] I have performed a self-review of my own code.

### CODE REVIEW

**Comment by sujeetsr** [comment]
Screenshot 2023-11-06 at 2 58 25 PM

File: src/http/data_loader.rs [hunk 1]
```diff
@@ -12,17 +13,37 @@ use crate::config::Batch;
...
 {
   pub client: C,
   pub batched: Option<GroupBy>,
+  #[allow(clippy::type_complexity)]
+  pub get_body_value: Arc<dyn Fn(HashMap<String, Vec<&ConstValue>>, &Cow<str>) -> ConstValue + Send + Sync>,
 }
 impl<C: HttpClient + Send + Sync + 'static + Clone> HttpDataLoader<C> {
-  pub fn new(client: C, batched: Option<GroupBy>) -> Self {
-    HttpDataLoader { client, batched }
+  pub fn new(client: C, batched: Option<GroupBy>, is_list: bool) -> Self {
```
Comment (amitksingh1490) [inline src/http/data_loader.rs:27]: can we separate into two Dataloaders

**Comment by sujeetsr** [comment]
Benchmark results with `ci-benchmark/benchmark.graphql` and `wrk -d 30 -t 4 -c 100 -s ci-benchmark/wrk.lua http://localhost:8000/graphql`
From main:
Screenshot 2023-11-08 at 10 09 58 AM

From this PR branch:
Screenshot 2023-11-08 at 10 14 29 AM

File: src/http/data_loader.rs [hunk 2]
```diff
@@ -16,10 +17,30 @@ use crate::json::JsonLike;
 pub struct HttpDataLoader {
   pub client: Arc<dyn HttpClient>,
   pub batched: Option<GroupBy>,
+  #[allow(clippy::type_complexity)]
+  pub get_body_value: Arc<dyn Fn(HashMap<String, Vec<&ConstValue>>, &Cow<str>) -> ConstValue + Send + Sync>,
```
Comment (meskill) [inline src/http/data_loader.rs:21]: it'd better to not supress clippy here but actually move part of the definition to separate type alias and reuse it

File: src/http/data_loader.rs [hunk 3]
```diff
@@ -16,10 +17,30 @@ use crate::json::JsonLike;
...
+            .collect::<Vec<_>>(),
+        )
+      },
+      false => |body_value: HashMap<String, Vec<&ConstValue>>, id: &Cow<str>| {
+        body_value
+          .get(id.as_ref())
+          .and_then(|a| a.first().cloned().cloned())
+          .unwrap_or(ConstValue::Null)
+      },
+    };
```
Comment (meskill) [inline src/http/data_loader.rs:42]: have you considered to store the flag itself HttpDataLoader and call just usual functions based on that flag?

I expect that approach be more readable that doesn't involve closures and arc wrapper

**Comment by sujeetsr** [comment]
Closing this, an alternative implementation with enum is done in https://github.com/tailcallhq/tailcall/pull/635

**Comment by sujeetsr** [comment]
Reopening, this, @meskill removed the Arc<dyn> from the get_body_value function, so there is no dynamic dispatch.

File: src/http/data_loader.rs [hunk 4]
```diff
@@ -156,4 +197,178 @@ mod tests {
...
+        .unwrap()
+        .as_ref()
+        .unwrap()
+        .as_ref()
+        .unwrap()
+        .body
+        .clone(),
+      ConstValue::Object(..)
+    ));
+  }
```
Comment (tusharmath) [inline src/http/data_loader.rs:373]: Can we write an `http_spec` using yml, instead of tests in .rs files.

File: src/http/data_loader.rs [hunk 5]
```diff
@@ -12,14 +12,42 @@ use crate::config::Batch;
...
+  pub get_body_value: fn(HashMap<String, Vec<&ConstValue>>, &str) -> ConstValue,
 }
 impl HttpDataLoader {
-  pub fn new(client: Arc<dyn HttpClient>, batched: Option<GroupBy>) -> Self {
-    HttpDataLoader { client, batched }
+  pub fn new(client: Arc<dyn HttpClient>, batched: Option<GroupBy>, is_list: bool) -> Self {
+    HttpDataLoader {
+      client,
+      batched,
+      get_body_value: if is_list {
```
Comment (tusharmath) [inline src/http/data_loader.rs:45]: #nitpick
```suggestion
      body: if is_list {
```

Comment (sujeetsr) [reply src/http/data_loader.rs]: Changed the field name

File: src/http/data_loader.rs (hunk 4 above)
Comment (sujeetsr) [reply src/http/data_loader.rs]: Updated `n-plus-one-list.yml` and removed the list tests in dataloader.

File: src/http/data_loader.rs [hunk 6]
```diff
@@ -156,4 +191,76 @@ mod tests {
...
+    );
+    assert_eq!(
+      client.requests.lock().unwrap().first().unwrap().url().to_string(),
+      "http://example.com/?userId=1"
+    );
+    assert_eq!(
+      client.requests.lock().unwrap().get(1).unwrap().url().to_string(),
+      "http://example.com/?userId=2"
+    );
+  }
```
Comment (tusharmath) [inline src/http/data_loader.rs:265]: Add the above tests into http spec. Delete theme from here.

**Review by tusharmath — APPROVED** [review]

**Pull request merged**